src/dune/fdo.ml

```diff
@@ -0,0 +1,213 @@
+open! Stdune
+module CC = Compilation_context
+
+type phase =
+  | All
+  | Compile
+  | Emit
+
+let linear_ext = ".cmir-linear"
+
+let linear_fdo_ext = linear_ext ^ "-fdo"
+
+let fdo_profile_filename s = s ^ ".fdo-profile"
+
+let linker_script_filename s = s ^ ".linker-script"
+
+let linker_script_hot_filename s = s ^ ".linker-script-hot"
+
+let phase_flags = function
+  | None -> []
+  | Some All -> [ "-g"; "-function-sections" ]
+  | Some Compile ->
+    [ "-g"; "-stop-after"; "scheduling"; "-save-ir-after"; "scheduling" ]
+  | Some Emit -> [ "-g"; "-start-from"; "emit"; "-function-sections" ]
+
+(* CR-soon gyorsh: add a rule to use profile with c/cxx profile if available,
+   similarly to opt_rule for ocaml modules. The profile will have to be
+   generated externally from perf data for example using google's autofdo
+   toolset: create_gcov for gcc or create_llvm_prof for llvm. *)
+let c_flags (ctx : Context.t) =
+  match ctx.fdo_target_exe with
+  | None -> []
+  | Some _ -> [ "-ffunction-sections" ]
+
+let cxx_flags = c_flags
+
+(* Location of ocamlfdo binary tool is independent of the module, but may
+   depend on the context. If it isn't cached elsewhere, we should do it here.
+   CR gyorsh: is it cached? *)
+let ocamlfdo_binary sctx dir =
+  let ocamlfdo =
+    Super_context.resolve_program sctx ~dir ~loc:None "ocamlfdo"
+      ~hint:"try: opam install ocamlfdo"
+  in
+  match ocamlfdo with
+  | Error e -> Action.Prog.Not_found.raise e
```

Comment:
Why should we raise here? It seems better to raise when a rule that requires this `ocamlfdo` is invoked instead.